File: cpp/questions/wordBreak.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int ss = s.size();
        std::vector<bool> dp(ss+1, false);
        dp[0] = true;
        for (int i=1; i<ss+1; ++i) {
            for (const auto& w: wordDict) {
                const int ws = w.size();
                const int sp = i - ws;
                if (sp>=0 && dp[sp] && s.substr(sp, ws)==w) {
                    dp[i] = true;
                }
            }
        }
        return dp[ss];
    }
};

File: python/questions/wordBreak.py
from typing import List

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        ls = len(s)
        dp = [False] * (ls + 1)
        dp[0] = True
        for i in range(1, ls+1):
            for w in wordDict:
                lw = len(w)
                sp = i - lw
                if sp>=0 and dp[sp] and s[sp:i]==w:
                    dp[i] = True
        return dp[ls]
